oak_containers_hello_world_untrusted_app/src/app_client.rs

```diff
@@ -25,41 +25,53 @@ mod proto {
 
 use anyhow::Context;
 use proto::oak::containers::example::{
-    untrusted_application_client::UntrustedApplicationClient as GrpcUntrustedApplicationClient,
+    trusted_application_client::TrustedApplicationClient as GrpcTrustedApplicationClient,
     HelloRequest,
 };
-use tokio_vsock::VsockStream;
 use tonic::transport::{Endpoint, Uri};
 use tower::service_fn;
 
 // Virtio VSOCK does not use URIs, hence this URI will never be used.
-// It is defined purely since in order to create a channel, since a URI has to
+// It is defined purely since in order to create a channel a URI has to
 // be supplied to create an `Endpoint`.
 static IGNORED_ENDPOINT_URI: &str = "file://[::]:0";
 
 /// Utility struct used to interface with the launcher
-pub struct UntrustedApplicationClient {
-    inner: GrpcUntrustedApplicationClient<tonic::transport::channel::Channel>,
+pub struct TrustedApplicationClient {
+    inner: GrpcTrustedApplicationClient<tonic::transport::channel::Channel>,
 }
 
-impl UntrustedApplicationClient {
-    pub async fn create(
-        untrusted_application_vsock_cid: u32,
-        untrusted_application_vsock_port: u32,
-    ) -> Result<Self, Box<dyn std::error::Error>> {
-        let inner: GrpcUntrustedApplicationClient<tonic::transport::channel::Channel> = {
+impl TrustedApplicationClient {
+    async fn get_stream_with_trusted_app(
+        cid: u32,
+        port: u32,
+    ) -> Result<tokio_vsock::VsockStream, anyhow::Error> {
+        let (vsock_stream, _) = tokio_vsock::VsockListener::bind(cid, port)
+            .context("failed to bind vsock listener")?
+            // The trusted app is the only party that will connect to this listener.
+            // Hence the first incoming stream must be the trusted app.
+            //
+            // Effectively this means that while on the gRPC layer the trusted app
+            // listens for invocations from the untrusted app, the inverse is
+            // true on the layer of the VSOCK connection. There the untrusted
+            // app listens for connections, the trusted app connects to the
+            // listener.
+            .accept()
+            .await
+            .context("failed to accept vsock connection")?;
+
+        Ok(vsock_stream)
+    }
```

Comment:
highlighting this for review, since it's a bit counter-intuitive. 

If we want to to avoid this, we could place the VSOCK listener in the trusted app and poll this listener on the untrusted side, waiting for the VM to boot up and the trusted app to start listening. 

I think that solution would be better than what is implemented in this PR. **However** in practice it proved difficult to implement, due to VSOCK exhibiting unexpected behavior. Specifically, other VSOCK connections in stage1 would fail if we started polling to early. Rather surprising, given that they take place on a different port.

Reply:
Given that we're considering pivoting from VSOCK to networking at the moment anyways, I think it makes sense to go ahead with the imperfect solution implemented here anyways. It'd unlock work on CI integration which is imo more urgent than debugging vsock.

Reply:
This is indeed counter-intuitive. I agree that the other way around will be better, but also agree the CI testing is more urgent. We can figure this out in a future PR.